### PR TITLE
Curves: Added serialization/ deserialization

### PR DESCRIPTION
**Pull request by Mugen87**
This PR adds serialization/ deserialization for all curve entities. It's now possible to store and load `ShapeGeometry` and `ShapeBufferGeometry`.

The implementation allows to perform serialization/ deserialization on all hierarchy levels (`Curve`, `CurvePath`, `Path`, `Shape`) with `.toJSON()` and `.fromJSON()` methods.

I'll consider `ExtrudeGeometry` in a separate PR.

Demo: https://jsfiddle.net/ht3zuzkr/2/

### CODE REVIEW

**Comment by Mugen87** [comment]
This PR directly addresses #8007

**Comment by mrdoob** [comment]
Hmm... It's kind of tempting to move `shapes` to its own group (along `geometries` and `materials` in the JSON format. In order to support this we'll probably need to add a UUID to `Shape` too. Would that make things too complicated?

**Comment by Mugen87** [comment]
Okay, I've changed the code so shapes are stored in their own group. `ShapeGeometry` and `ShapeBufferGeometry` now refer with a `uuid` to their shapes.

Updated Demo: https://jsfiddle.net/ht3zuzkr/2/

**Comment by Mugen87** [comment]
We could actually also do this for all `Curve` types (so a `Shape` object refers to their curves with `uuid`s) but i have the feeling this goes a little bit too far.

**Comment by Mugen87** [comment]
@mrdoob Is the PR now okay like this? 😊

**Comment by mrdoob** [comment]
Checking!

File: src/core/BufferGeometry.js [hunk 1]
```diff
@@ -923,6 +923,32 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
 
 			}
 
+			// ShapeBufferGeometry
+
+			if ( data.type === 'ShapeBufferGeometry' ) {
```
Comment (mrdoob) [inline src/core/BufferGeometry.js]: How about adding this in `ShapeBufferGeometry`'s own `toJSON()`?

```js
ShapeBufferGeometry.prototype.toJSON = function () {

	var data = BufferGeometry.prototype.toJSON.call( this );

	data.shapes = [];

	if ( Array.isArray( shapes ) ) {

		for ( var i = 0, l = shapes.length; i < l; i ++ ) {

			var shape = shapes[ i ];

			data.shapes.push( shape.uuid );

		}

	} else {

		data.shapes.push( shapes.uuid );

	}

	return data;

}
```

I think that should work?

File: src/core/Geometry.js [hunk 2]
```diff
@@ -1009,6 +1009,32 @@ Object.assign( Geometry.prototype, EventDispatcher.prototype, {
 
 			}
 
+			// ShapeGeometry
+
+			if ( data.type === 'ShapeGeometry' ) {
```
Comment (mrdoob) [inline src/core/Geometry.js]: Same here...

**Comment by mrdoob** [comment]
Apart from these two. It's looking good to me!

File: src/core/BufferGeometry.js (hunk 1 above)
Comment (Mugen87) [reply src/core/BufferGeometry.js]: Yes. Indeed, it's better to place this code not in `Geometry` or `BufferGeometry` but in `ShapeGeometry`.

**Comment by Mugen87** [comment]
Updated Demo https://jsfiddle.net/ht3zuzkr/2/

File: src/core/Object3D.js [hunk 3]
```diff
@@ -670,6 +671,31 @@ Object.assign( Object3D.prototype, EventDispatcher.prototype, {
 
 			object.geometry = serialize( meta.geometries, this.geometry );
 
+			var parameters = this.geometry.parameters;
+
+			if ( parameters !== undefined && parameters.shapes !== undefined ) {
+
+				var shapes = parameters.shapes;
+				var uuids = [];
```
Comment (mrdoob) [inline src/core/Object3D.js]: What's this array for? 🤔

Comment (Mugen87) [reply src/core/Object3D.js]: Oh, this is a remnant of a previous solution. I'll remove it...

**Comment by mrdoob** [comment]
Beautiful! 😊

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**